### DISCUSSION TITLE
smtpd processes becoming uninterruptible.

### DISCUSSION THREAD

**Discussion by grheard** [docker-mailserver/docker-mailserver#3697] [Q&A]
I added docker-mailserver to my unraid server some time back and everything was fine for a while, but lately some hacking attempts on port 25 have been leading to smptd processes becoming stuck in an uninterruptible state. When this happens the unraid server eventually reaches a high load and other docker containers begin to fail to respond. Attempting to stop the mailserver docker fails and the unraid array cannot be stopped cleanly leading to a lenghtly parity check on restart. Once its back up again its usually only a day or two before another round of the same type of hacking attempt is made which leads to the same outcome for the server. Without mailserver running the unraid server is stable and does not suffer from these periodic 'hangs'.

Here is the pattern seen in the logs that leads to the smtpd processes becoming hung:
```
Nov 30 10:13:18 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:12391 to [172.17.0.7]:25
Nov 30 10:13:18 mail postfix/postscreen[4329]: HANGUP after 0 from [91.92.248.232]:12391 in tests before SMTP handshake
Nov 30 10:13:18 mail postfix/postscreen[4329]: DISCONNECT [91.92.248.232]:12391
Nov 30 10:13:18 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:12783 to [172.17.0.7]:25
Nov 30 10:13:24 mail postfix/postscreen[4329]: PASS NEW [91.92.248.232]:12783
Nov 30 10:13:26 mail postfix/smtpd[4339]: connect from unknown[91.92.248.232]
Nov 30 10:13:26 mail postfix/smtpd[4339]: NOQUEUE: reject: RCPT from unknown[91.92.248.232]: 554 5.7.1 <test@gmail.com>: Relay access denied; from=<test@sbcglobal.net> to=<test@gmail.com> proto=SMTP helo=<win-clj1b0gq6jp.domain>
Nov 30 10:13:26 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:15929 to [172.17.0.7]:25
Nov 30 10:13:26 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:15929
Nov 30 10:13:26 mail postfix/smtpd[4346]: connect from unknown[91.92.248.232]
Nov 30 10:13:27 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16043 to [172.17.0.7]:25
Nov 30 10:13:27 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16043
Nov 30 10:13:27 mail postfix/smtpd[4347]: connect from unknown[91.92.248.232]
Nov 30 10:13:27 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16190 to [172.17.0.7]:25
Nov 30 10:13:27 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16190
Nov 30 10:13:27 mail postfix/smtpd[4351]: connect from unknown[91.92.248.232]
Nov 30 10:13:27 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16322 to [172.17.0.7]:25
Nov 30 10:13:27 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16322
Nov 30 10:13:27 mail postfix/smtpd[4352]: connect from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16525 to [172.17.0.7]:25
Nov 30 10:13:28 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16525
Nov 30 10:13:28 mail postfix/smtpd[4353]: connect from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16638 to [172.17.0.7]:25
Nov 30 10:13:28 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16638
Nov 30 10:13:28 mail postfix/smtpd[4354]: connect from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/smtpd[4347]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/smtpd[4346]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/smtpd[4347]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:28 mail postfix/smtpd[4346]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:28 mail postfix/smtpd[4351]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/smtpd[4351]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:28 mail postfix/smtpd[4339]: lost connection after RCPT from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/smtpd[4339]: disconnect from unknown[91.92.248.232] helo=1 mail=1 rcpt=0/1 commands=2/3
Nov 30 10:13:28 mail postfix/smtpd[4352]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:28 mail postfix/smtpd[4352]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:28 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16810 to [172.17.0.7]:25
Nov 30 10:13:28 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16810
Nov 30 10:13:28 mail postfix/smtpd[4347]: connect from unknown[91.92.248.232]
Nov 30 10:13:29 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:16877 to [172.17.0.7]:25
Nov 30 10:13:29 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:16877
Nov 30 10:13:29 mail postfix/smtpd[4346]: connect from unknown[91.92.248.232]
Nov 30 10:13:29 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:17032 to [172.17.0.7]:25
Nov 30 10:13:29 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:17032
Nov 30 10:13:29 mail postfix/smtpd[4351]: connect from unknown[91.92.248.232]
Nov 30 10:13:30 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:17165 to [172.17.0.7]:25
Nov 30 10:13:30 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:17165
Nov 30 10:13:30 mail postfix/smtpd[4339]: connect from unknown[91.92.248.232]
Nov 30 10:13:30 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:17458 to [172.17.0.7]:25
Nov 30 10:13:30 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:17458
Nov 30 10:13:30 mail postfix/smtpd[4352]: connect from unknown[91.92.248.232]
Nov 30 10:13:30 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:17654 to [172.17.0.7]:25
Nov 30 10:13:30 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:17654
Nov 30 10:13:30 mail postfix/smtpd[4358]: connect from unknown[91.92.248.232]
Nov 30 10:13:31 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:17790 to [172.17.0.7]:25
Nov 30 10:13:31 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:17790
Nov 30 10:13:31 mail postfix/smtpd[4359]: connect from unknown[91.92.248.232]
Nov 30 10:13:31 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:17947 to [172.17.0.7]:25
Nov 30 10:13:31 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:17947
Nov 30 10:13:31 mail postfix/smtpd[4363]: connect from unknown[91.92.248.232]
Nov 30 10:13:31 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:18112 to [172.17.0.7]:25
Nov 30 10:13:31 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:18112
Nov 30 10:13:31 mail postfix/smtpd[4364]: connect from unknown[91.92.248.232]
Nov 30 10:13:32 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:18275 to [172.17.0.7]:25
Nov 30 10:13:32 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:18275
Nov 30 10:13:32 mail postfix/smtpd[4365]: connect from unknown[91.92.248.232]
Nov 30 10:13:32 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:18410 to [172.17.0.7]:25
Nov 30 10:13:32 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:18410
Nov 30 10:13:32 mail postfix/smtpd[4366]: connect from unknown[91.92.248.232]
Nov 30 10:13:32 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:18644 to [172.17.0.7]:25
Nov 30 10:13:32 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:18644
Nov 30 10:13:32 mail postfix/smtpd[4367]: connect from unknown[91.92.248.232]
Nov 30 10:13:33 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:18770 to [172.17.0.7]:25
Nov 30 10:13:33 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:18770
Nov 30 10:13:33 mail postfix/smtpd[4368]: connect from unknown[91.92.248.232]
Nov 30 10:13:33 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:18921 to [172.17.0.7]:25
Nov 30 10:13:33 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:18921
Nov 30 10:13:33 mail postfix/smtpd[4372]: connect from unknown[91.92.248.232]
Nov 30 10:13:33 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:19081 to [172.17.0.7]:25
Nov 30 10:13:33 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:19081
Nov 30 10:13:33 mail postfix/smtpd[4373]: connect from unknown[91.92.248.232]
Nov 30 10:13:34 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:19195 to [172.17.0.7]:25
Nov 30 10:13:34 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:19195
Nov 30 10:13:34 mail postfix/smtpd[4374]: connect from unknown[91.92.248.232]
Nov 30 10:13:34 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:19317 to [172.17.0.7]:25
Nov 30 10:13:34 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:19317
Nov 30 10:13:34 mail postfix/smtpd[4375]: connect from unknown[91.92.248.232]
Nov 30 10:13:34 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:19582 to [172.17.0.7]:25
Nov 30 10:13:34 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:19582
Nov 30 10:13:34 mail postfix/smtpd[4376]: connect from unknown[91.92.248.232]
Nov 30 10:13:35 mail postfix/postscreen[4329]: CONNECT from [91.92.248.232]:19719 to [172.17.0.7]:25
Nov 30 10:13:35 mail postfix/postscreen[4329]: PASS OLD [91.92.248.232]:19719
Nov 30 10:13:35 mail postfix/smtpd[4377]: connect from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4366]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4366]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4376]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4376]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4377]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4377]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4365]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4365]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4373]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4364]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4359]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4373]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4372]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4364]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4359]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4372]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4363]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4363]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4339]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4358]: lost connection after EHLO from unknown[91.92.248.232]
Nov 30 10:13:40 mail postfix/smtpd[4358]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
Nov 30 10:13:40 mail postfix/smtpd[4339]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
```

These are the last lines before the docker becomes unresponsive.

As I am not seeing people talking about this kind of problem, I am pretty sure there is something in how I configured the docker that may be leaving it open to this vulnerability.

Here is the docker run command used:
```
docker run
  -d
  --name='mailserver'
  --net='bridge'
  -e TZ="America/New_York"
  -e HOST_OS="Unraid"
  -e HOST_HOSTNAME="****"
  -e HOST_CONTAINERNAME="mailserver"
  -e 'ENABLE_SPAMASSASSIN'='1'
  -e 'SPAMASSASSIN_SPAM_TO_INBOX'='1'
  -e 'ENABLE_CLAMAV'='1'
  -e 'ENABLE_FAIL2BAN'='1'
  -e 'ENABLE_POSTGREY'='1'
  -e 'ENABLE_SASLAUTHD'='0'
  -e 'ONE_DIR'='1'
  -e 'SSL_TYPE'='letsencrypt'
  -l net.unraid.docker.managed=dockerman
  -l net.unraid.docker.icon='https://docker-mailserver.github.io/docker-mailserver/v11.1/assets/logo/dmo-logo-white.min.svg'
  -p '25:25/tcp'
  -p '143:143/tcp'
  -p '587:587/tcp'
  -p '993:993/tcp'
  -p '465:465/tcp'
  -v '/mnt/user/appdata/mailserver/var/mail':'/var/mail':'rw'
  -v '/mnt/user/appdata/mailserver/var/mail-state':'/var/mail-state':'rw'
  -v '/mnt/user/appdata/mailserver/var/log/mail':'/var/log/mail':'rw'
  -v '/mnt/user/appdata/mailserver/tmp/docker-mailserver':'/tmp/docker-mailserver':'rw'
  -v '/etc/localtime':'/etc/localtime':'ro'
  -v '/mnt/user/appdata/swag/etc/letsencrypt':'/etc/letsencrypt':'ro'
  --hostname=mail
  --domainname=*******.***
  --cap-add=NET_ADMIN 'mailserver/docker-mailserver:latest'
```

Any help/suggestions as to how to fix or diagnose the issue would be greatly appreciated.

**Comment by polarathene** [comment]
> ```
> --hostname=mail
> --domainname=*******.***
> ```

Probably won't make a difference but that should all be in `--hostname`, `--domainname` is for NIS domain which is different.

---

> `91.92.248.232`

How often is the IP changing here in the logs?

If it's not changing, you could block it.

Otherwise there might be some additional config to identify the behaviour and block with Fail2Ban, or more strict restrictions on Postfix for what connections are permitted, not sure if that'd help here with that kind of activity though.

---

> `-l net.unraid.docker.icon='https://docker-mailserver.github.io/docker-mailserver/v11.1/assets/logo/dmo-logo-white.min.svg'`

This is all I have to go by, but are you actually on DMS v11.1 still? Upgrading might help, we're at v13 now. Many fixes + updates since.

---

> ```
> Nov 30 10:13:28 mail postfix/smtpd[4346]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
> Nov 30 10:13:28 mail postfix/smtpd[4351]: lost connection after EHLO from unknown[91.92.248.232]
> Nov 30 10:13:28 mail postfix/smtpd[4351]: disconnect from unknown[91.92.248.232] ehlo=1 commands=1
> Nov 30 10:13:28 mail postfix/smtpd[4339]: lost connection after RCPT from unknown[91.92.248.232]
> Nov 30 10:13:28 mail postfix/smtpd[4339]: disconnect from unknown[91.92.248.232] helo=1 mail=1 rcpt=0/1 commands=2/3
> ```

I am curious if enabling [`POSTFIX_REJECT_UNKNOWN_CLIENT_HOSTNAME=1`](https://docker-mailserver.github.io/docker-mailserver/v13.0/config/environment/#postfix_reject_unknown_client_hostname) will work well for you? 

[Postfix docs](https://www.postfix.org/postconf.5.html#reject_unknown_client_hostname) (_I've slightly modified it in the quote_):

> Reject the request when any of these conditions are true:
> 1) the client IP address->name mapping fails
> 2) the name->address mapping fails
> 3) the name->address mapping does not match the client IP address.
> 
> This is a stronger restriction than the [`reject_unknown_reverse_client_hostname`](https://www.postfix.org/postconf.5.html#reject_unknown_reverse_client_hostname) feature, which triggers only under condition 1) above.

This was suggested recently [for this discussion](https://github.com/orgs/docker-mailserver/discussions/3687#discussioncomment-7795743), but their issue was related to a proxy routing their ingress from the client to DMS which lost the source IP address, and the proxy IP that replaced it was considered trusted.

The feature was temporarily enabled as an improvement to block spam, but some users found it was a bit too aggressive for them so we've required it to be opt-in.

`reject_unknown_reverse_client_hostname` is more relaxed as the quoted docs mention, so if it is too aggressive for you, you could instead try setting this via [our `user-patches.sh` support](https://docker-mailserver.github.io/docker-mailserver/v13.0/config/advanced/override-defaults/user-patches/). It'd look similar to how the mentioned ENV feature works:

```bash
#!/bin/bash

echo 'Enabling "reject_unknown_reverse_client_hostname" to "dms_smtpd_sender_restrictions"' 
sedfile -i -E \ 
    's|^(dms_smtpd_sender_restrictions = .*)|\1, reject_unknown_reverse_client_hostname|' \ 
    /etc/postfix/main.cf 
```

**Reply by grheard** [reply]
Thanks for the reply!

> This is all I have to go by, but are you actually on DMS v11.1 still?

I am running the latest v13. That old url is where the unraid docker client is pulling the projects logo from.

> How often is the IP changing here in the logs?
> 
> If it's not changing, you could block it.
> 
> Otherwise there might be some additional config to identify the behavior and block with Fail2Ban, or more strict restrictions on Postfix for what connections are permitted, not sure if that'd help here with that kind of activity though.

I was curious as to why fail2ban did not act in this instance, but maybe it was because the connections were not dropped until later to be considered for a ban? I'm still looking through its documentation.

I have had to manually add a couple international IP domains because they were repeatedly attempting connections every few seconds from differing IPs within the same subnet. Since that worked, I do know that fail2ban is actually capable of blocking.

> I am curious if enabling [POSTFIX_REJECT_UNKNOWN_CLIENT_HOSTNAME=1](https://docker-mailserver.github.io/docker-mailserver/v13.0/config/environment/#postfix_reject_unknown_client_hostname) will work well for you?

> This was suggested recently [for this discussion](https://github.com/orgs/docker-mailserver/discussions/3687#discussioncomment-7795743), but their issue was related to a proxy routing their ingress from the client to DMS which lost the source IP address, and the proxy IP that replaced it was considered trusted.
> 
> The feature was temporarily enabled as an improvement to block spam, but some users found it was a bit too aggressive for them so we've required it to be opt-in.
> 
> reject_unknown_reverse_client_hostname is more relaxed as the quoted docs mention, so if it is too aggressive for you, you could instead try setting this via [our user-patches.sh support](https://docker-mailserver.github.io/docker-mailserver/v13.0/config/advanced/override-defaults/user-patches/). It'd look similar to how the mentioned ENV feature works:

I will have a look at these as possible solutions. In the 3 instances I have archived of this same behavior, none of them even succeeded in the name lookup so reject_unknown_reverse_client_hostname may be enough to stop these kinds of 'attacks'.

At the moment I am test driving the following changes in postfix-main.cf:

```
smtpd_client_connection_count_limit = 3
smtpd_client_connection_rate_limit = 3
```

Legitimate clients shouldn't need very many connections or to connect that often. I thought this could curtail the rapid activity as portrayed in the log snippet. If I have read the documentation right this allows the same client ip to connect a maximum of 3 times (concurrently or in a row) within the anvil time limit which is currently the default of 60s. So the same client should be able to connect a maximum of 3 times in 60s.

All of my internal services use the 465 submission port to send email so this should only impact external servers attempting to deliver email to my domain.

Thanks again for the pointers. I will continue to dig through the dms and postfix documentation.
